Upgrade to Support hyper-0.10

jolhoeft opened the pull request.

jolhoeft committed:
Upgrade to support nickel 0.9

jolhoeft committed:
Dependemcy upgrades

jolhoeft committed:
Merge remote-tracking branch 'upstream/master'

jolhoeft committed:
Upgrade to use nickel 0.10

Ryman commented on src/cookies.rs:
```diff
@@ -26,17 +26,18 @@ impl SecretKey {
-            Some(c) => c.to_cookie_jar(&key.0),
-            None => CookieJar::new(&key.0)
+            //Some(c) => c.to_cookie_jar(&key.0),
+            Some(c) => c.iter().filter_map(|s| s.parse::<Cookie>().ok()).
```
Can you remove the commented old code and knock this into it's own `{ block }` for ease of reading?

Ryman replied:
Also maybe a note that malformed cookies are dropped here in the filter. (Seems fine to do that, perhaps it could go into a `debug!` log or something though incase someone did need to debug where their cookies went)

jolhoeft replied:
Formatting changes in latest version of PR

Ryman commented on src/cookies.rs:
```diff
@@ -47,18 +48,18 @@ impl<'a, 'b, 'k, D> Plugin<Response<'a, D>> for CookiePlugin
 
-        let key = res.server_data().key();
-        Ok(CookieJar::new(&key.0))
+        Ok(CookieJar::new())
```
Does the new CookieJar no longer support encryption? I feel like if we're to drop that as a feature then we should at least change some of the documentation around this library or implement those features using another library, perhaps behind a feature?

jolhoeft replied:
I don't fully recall now why I had to remove this. Looking at the docs for the cookie crate, it looks like signed cookies are supported, but the api has changed. When I made this patch, I was trying to get our project working with nickel-0.10. We aren't using encrypted cookies, so I punted when it turned complicated.

It may be best to regard this pull request as more documentation of some of the issues I discovered.

jolhoeft replied:
Done in latest version of PR

jolhoeft replied:
I've had more time to look into this. The API changed from 0.2 to 0.8 and on. Now CookieJar is not encrypted or signed, but cookies in the jar can be by the private and signed methods of CookieJar. Basically you would use cookies_mut to get a mutable jar, and use those methods for secure access. I believe we don't need to do anything more to support secure cookies.

jolhoeft replied:
My previous comment only applies to 0.10. 0.8 doesn't have this feature, and signing/encryption is on a per cookie basis.

I am working on updating the example. At the same time I think I should update to cookie-0.10. Your thoughts?

jolhoeft replied:
I was wrong, 0.8 supports this feature. I have updated the example to use the cookie crate's new approach to secure access. In the process I removed the AllowMutCookies trait, since the secured access requires mutable access to the Request cookies.

There might be some room for a cleaner API, but I think that ought to be part of a different PR.

Ryman reviewed (changes requested):
Hey, I have a concern about the implicit dropping of encrypted cookie support here unless I'm missing something. Do you know if that is still an option with the newer version of cookies?

Also, would be good to squash it down to a single commit but it looks like you can do that on the github UI now so no need to worry until we're ready to merge!

jolhoeft committed:
Some requested format changes

jolhoeft committed:
Restore signed and secured cookies

This is mostly a change in the example. I removed the
AllowMutCookies trait since the cookies crate now requires mutable
access to the Request cookies for secured access.

jolhoeft commented:
See my comments above. This may be ready to be merged.

jolhoeft commented:
Any thoughts?

jolhoeft committed:
Update to latest cookie crate

Ryman commented on examples/cookies_example.rs:
```diff
@@ -2,30 +2,55 @@
+fn secure_middleware<'mw, 'conn, D: KeyProvider>(req: &mut Request<'mw, 'conn, D>, mut res: Response<'mw, D>) -> MiddlewareResult<'mw, D> {
+    let old_value = { // block for borrow management
+        let server_key = res.server_data().key();
+        let key = Key::from_master(&server_key.0);
```
Would it make sense to change the return type from `.key()` to what cookie requires here?

jolhoeft replied:
Change it from `SecretKey` to `[u8; 32]`? That might work. Would that mean `SecretKey` no longer needs to be public? That would reduce the chance of name-space collision, which seems like a small win too.

Ryman replied:
I was thinking that you could change the return type to `cookies::Key` itself? Is that already `[u8; 32]`?

jolhoeft replied:
`cookie::Key` does not exist if the secure feature is not enabled. We might able to work around that, but that looks to be some fairly extensive changes I'd prefer to bundle into a different PR.

Ryman commented on src/cookies.rs:
```diff
@@ -96,7 +92,7 @@ pub trait KeyProvider {
 
 #[cfg(feature = "secure")]
-impl KeyProvider for () {}
+impl<T> KeyProvider for T {}
```
I'm unsure about this, the original idea if I remember was that if `secure` was active then the user has to implement this (by providing their own key) or use the `()` implementation to have a random key per server restart. With this change if I understand it correctly the users will by default get a random key and be unable to implement the trait for their own type without specialization?

jolhoeft replied:
I see what you mean. Since we provide a default implementation for the trait, do we need to provide anything for this case? Providing an implementation for `()` requires the user to create a server data struct if they want a non-random key.

This trait cause me some confusion when I was first using nickel-cookies, because it requires my server data to implement KeyProvider, but that wasn't obvious. That was my rational for implementing it for everything. I'm not quite sure how to clarify that in the documentation, though.

Ryman replied:
If we don't supply the implementation for `()` (the default ServerData type) then the user would have to use a custom type and implement this trait if they want to use cookies I think. And they won't be able to implement it themselves for `()` due to coherence issues.

Yeah, the documentation on this is very poor, which is I guess an artifact of it being kind of an experiment that got used :P

jolhoeft replied:
Oh right. The means you can't use anything you didn't define yourself as server data. Annoying, but easy enough to work around. I will revert this.

Ryman commented:
@jolhoeft Sorry for the delay, glad to hear the encryption is still supported. I just have a concern on the blanket impl for the KeyProvider trait but other than that I think it looks good 👍

jolhoeft committed:
Fix example to work with --no-default-features

jolhoeft committed:
Revert overly general trait implementation

jolhoeft commented:
I've removed the blanket impl for the KeyProvider. How does it look now?

Ryman commented on src/cookies.rs:
```diff
@@ -47,25 +50,21 @@ impl<'a, 'b, 'k, D> Plugin<Response<'a, D>> for CookiePlugin
             let header = {
                 let jar = res.get_ref::<CookiePlugin>().unwrap();
-                header::SetCookie::from_cookie_jar(jar)
+                // header::SetCookie::from_cookie_jar(jar)
```
nit: commented dead code

Ryman commented:
LGTM but I haven't seen the code executed so take my opinion with a grain of salt :D

eltomito commented:
Could you please merge this pull request?

jolhoeft commented:
Sorry, I let this drop off my radar.

jolhoeft commented:
I've merged this into master, but I do not have the necessary permissions to push this to crates.io. @Ryman could you take care of that? You are the only person who can push the crate.

Ryman commented:
@jolhoeft sorry for the delay, added you to the owners for the crate!
